Added signal to NavigationAgent when entering a link

DarkKilauea opened the pull request:
Finishes resolving godotengine/godot-proposals#2527

Depends on: #66729 (that PR is included in a separate commit)

This PR finally adds support for signaling on the `NavigationAgent` when it has reached the start of a navigation link.  This allows users to perform custom movement logic at that point.  I also extended the information provided by `NavigationPathQueryResult` to include metadata about each point in the path so custom implementations can act upon the regions and links entered along the way.

Also exposed the underlying query result from `NavigationAgent` so folks using it have access to the full metadata about the path.

Open Questions:
1. Does anyone have opinions on the arguments in the new signals?  I tried to include what was helpful during testing, but some community feedback would be helpful.

DarkKilauea committed:
Add support for emitting a signal when entering a NavLink

DarkKilauea commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -228,7 +228,11 @@ Vector<Vector3> GodotNavigationServer::map_get_path(RID p_map, Vector3 p_origin,
+	TypedArray<RID> path_rids = TypedArray<RID>();
+	Vector<int64_t> path_owners = Vector<int64_t>();
+
+	return map->get_path(p_origin, p_destination, p_optimize, p_navigation_layers, path_types, path_rids, path_owners);
```
Right now I'm using references here, which avoids the chance of passing `nullptr` and having to sprinkle `if` statements throughout `NavMap::get_path`.  However, it does force me to collect the metadata here for the "legacy" `map_get_path` implementation to avoid breaking compatibility, only to throw that metadata away.

Those with more experience in C++, would it be better to pass `nullptr` here and add checks throughout `NavMap::get_path`?  Are there any clever compiler optimizations that might influence the approach I should take here?

smix8 replied:
Don't upgrade / support the old `map_get_path()` function, leave it alone, it is not worth it.

DarkKilauea replied:
No user facing changes have been made, but I did go ahead and convert the references to pointers in order to avoid any allocation overhead due to metadata collection.

This method should now perform basically the same as it did before (only theoretical cost would be some branch instructions, which the compiler will probably optimize out).

akien-mga commented:
CC @smix8 @lawnjelly

smix8 commented:
This additional info should be made optional as a pathquery parameter and branched inside the server. Also don't upgrade / support the old `map_get_path()` function, leave it alone, it is not worth it.

One of the main reasons for the pathquery and pathresult was that we can change parameters without breaking everything like an exposed function would and to branch inside the internal functions depending on what the user wants. Avoid to tinker with the current main pathfinding function. Create a new function if you want to support an optional pathquery parameter that requires branching. E.g. if the user has parameter `use_navlinks = false` or `get_region_ids = false` it makes no sense to do all the extra work in every single path query.

Yes this may create situations where we need to duplicated some pathfinding code due to how the loops work but the current one is already hard enough to maintain, adding a ton of extra features to a giant get_path() function will paint us against a performance and maintainability wall at some point.

smix8 commented:
@venilark
This is an extention of the navlink pr with the part that could not be done before the pathquery pr was done and merged so it will be added when it is done and tested. The navlinks have the same `navigation_layers` bitmask as navregions as both share the same base.

DarkKilauea commented:
I'm thinking an overall rework of the pathing logic should be done in another PR.  I'm hesitant to duplicate the logic, knowing that we are already talking about reworking it in general (plus that duplication would complicate review of these changes).  I think the current changes are relatively unobtrusive, but I agree with you that we shouldn't add much more without a rework.

DarkKilauea commented:
Ok, I added a `path_metadata_flags` member to the query and agent which controls the additional information that is fetched during the path query itself.  This should resolve @smix8 concerns about doing unnecessary additional processing that the user did not need.

I also changed the signals to return dictionaries with the information about the waypoint or link that was hit.

In the UI, `NavigationAgent` now looks like:
![image](https://user-images.githubusercontent.com/1929107/194959204-bafec050-5375-44c9-ab52-9a3808cbd7b7.png)

venilark commented:
Has an autojump feature been considered, similar to Unreal Engine's?
![autojump](https://user-images.githubusercontent.com/98270528/195983704-24e4db79-84d5-43ff-8e34-924ea00b8351.PNG)
in case we don't need ladders or more complex stuff so the Agents can easily traverse simple levels with just platforms without custom solutions (now we'll get the signal which is already a huge improvement but the movement logic has to be designed manually)

smix8 commented:
@venilark
No, the entire NavigationAgent movement logic is designed to be manually done with a script, the link handling is no exception.

Does not prevent anyone from sharing a NavigationAgent script template that supports movement including links.

venilark commented:
@smix8 
in this case we need to manually place the Path3D in the same exact position as both links and the signal should return the ID of that link or something like that, so we know which end of the Path we have to place the Agent on and start the jumping animation
question: if we do that, wouldn't we get a second signal once the Agent lands on the second link, or the signal gets triggered only when actually using the link, instead of stepping on it?

DarkKilauea commented:
@venilark 
With the current implementation in this PR, you would get a signal for starting a link, then another signal for reaching the next way point.

The signal is only triggered when actually navigating using a link.  Passing by a link without the path including it would not trigger the signal.

Both signals in the PR give you a reference to the node which manages the link or region which contains the way point, so you should be able to write a script to add additional details to the link (like path to follow, animation to play, etc).  You should also be able to write an EditorPlugin to add gizmo support for your added gameplay logic so you can see it in editor.

I don't intend to add built-in support for movement, since the goal of Godot is to be a general purpose engine that avoids prescribing a way to do things (when feasible).  Unreal is a bit different, in that is was built to be a FPS engine first, so it contains some assumptions about the game following an FPS design.

venilark commented:
@DarkKilauea 

I understand, my suggestion was simply to add the in-engine possibility to make the agents jump automatically like UE and Unity do, completely optional for people who only care about jumping (no ladders, no ledge grabbing..), something they could see in a youtube tutorial and have test capsules traversing the map in minutes.

venilark commented:
About the parameters in the signal, imagine I add a Path3D in the same exact location as the link to handle the jump, you return the location of the link being used but then I need to loop through all my Path3Ds to get the one that has that exact Vector3 as the starting point, am I wrong? If this is the case, a custom string id in the link itself that gets returned in the signal maybe could help.

DarkKilauea commented:
@venilark 

In that case, I would add a script to `NavigationLink3D` along the lines of:
``` gdscript
class_name MyJumpLink
extends NavigationLink3D

@export var path: NodePath;
@onready var path_node: Path3D = get_node(path) as Path3D;
```

Then, inside your agent, you can pull the link out of the event, get a reference to your link, and pull out the path.  Something like:

``` gdscript
func on_link_reached(details: Dictionary):
    var link: MyJumpLink = details["owner"] as MyJumpLink;
    var path := link.path_node;

    # Do what you need to do to follow the path.
```

Note: I've written the scripts from memory, so their may be typos, but the general idea should be sound.  This is the nice thing about returning the node that "owns" the link, it allows you to add scripts or derive from it and add whatever you need to.

akien-mga commented on doc/classes/NavigationAgent2D.xml:
```diff
@@ -137,6 +146,17 @@
+				Notifies when a navigation link has been reached.
+				The details dictionary may contain the following keys depending on the value of [member path_metadata_flags]:
+				- location: The start location of the link that was reached.
+				- type: Always [code]NavigationPathQueryResult2D.PATH_SEGMENT_TYPE_LINK[/code].
```
```suggestion
				- type: Always [constant NavigationPathQueryResult2D.PATH_SEGMENT_TYPE_LINK].
```

akien-mga replied:
Maybe also make the keys use `[code]` formatting in each of the descriptions of dictionary keys, so it's clear what they are (and so translators don't think they're English words and translate them).

DarkKilauea replied:
Fixed.

akien-mga commented on doc/classes/NavigationAgent3D.xml:
```diff
@@ -143,6 +152,17 @@
+				Notifies when a navigation link has been reached.
+				The details dictionary may contain the following keys depending on the value of [member path_metadata_flags]:
+				- location: The start location of the link that was reached.
+				- type: Always [code]NavigationPathQueryResult3D.PATH_SEGMENT_TYPE_LINK[/code].
```
```suggestion
				- type: Always [constant NavigationPathQueryResult3D.PATH_SEGMENT_TYPE_LINK].
```

DarkKilauea replied:
Fixed.

DarkKilauea commented:
Rebased on current `master` branch.

akien-mga commented:
Thanks!

venilark commented:
When setting up the NavigationLink3D, start position and end position are in relation to the NavigationLink3D itself, not global coordinates. When we get the details in the signal, we get the global coordinates of the current part of the link we are stepping on, whether it's start pos or end pos, so how can we know which one we are on?
Could the signal return a string like "link_position_reached": "start"/"end" and ?

Also the documentation says
- location: The start location of the link that was reached.
This is confusing, because if a link is bidirectional, you could be on the "end"

venilark commented:
![j1](https://user-images.githubusercontent.com/98270528/212495312-b2f7c85b-6003-46ba-95c1-0e1efecb6e89.PNG)
@DarkKilauea I came up with this idea and found out what comes in the Details dictionary and the Y position of the node are different, probably because it returns the closest point intersecting the Navmesh. Is this the intended behaviour?
The code returns -1 because it is not exact enough so then I did this
if current_node_position.distance_to(start_node_position) < 1:
		return 0
	elif current_node_position.distance_to(end_node_position) < 1:
		return 1
	return -1
	
it works but seems a bit hacky
